packages/rich-list-core/src/RichListCore.ts

```diff
@@ -0,0 +1,219 @@
+import { ApiClient, BigNumber, blockchain as defid } from '@defichain/jellyfish-api-core'
+import { QueueClient, Queue } from './lib/Queue'
+import { SingleIndexDb, Schema } from './lib/SingleIndexDb'
+import { AddressParser } from './controller/AddressParser'
+import { NetworkName } from '@defichain/jellyfish-network'
+import { RichListItem } from '@defichain/rich-list-api-client'
+import { AccountAmount } from '@defichain/jellyfish-api-core/src/category/account'
+import { ActiveAddressAccountAmount } from './controller/AddressParser/ActiveAddressAccountAmount'
+
+const DEFAULT_RICH_LIST_LENGTH = 1000
+
+export class RichListCore {
+  isCatchingUp = false
+  richListLength = DEFAULT_RICH_LIST_LENGTH
+  readonly addressParser: AddressParser
+
+  constructor (
+    private readonly network: NetworkName,
+    private readonly apiClient: ApiClient,
+    private readonly existingRichList: SingleIndexDb<RichListItem>,
+    private readonly crawledBlockHashes: SingleIndexDb<CrawledBlock>,
+    private readonly droppedFromRichList: SingleIndexDb<string>,
+    private readonly queueClient: QueueClient<string>
+  ) {
+    this.addressParser = new AddressParser(apiClient, network)
+  }
+
+  setRichListLength (length: number): void {
+    this.richListLength = length
+  }
+
+  resume (): void {
+    if (this.isCatchingUp) {
+      return
+    }
+    this.isCatchingUp = true
+    void this._catchUp()
+  }
+
+  private async _catchUp (): Promise<void> {
+    const lastBlock = await this._getCrawledTip()
+    const nextBlockHeight = lastBlock?.data.height ?? 0
+    const nextBlock = await this._getBlock(nextBlockHeight)
+
+    if (nextBlock === undefined) {
+      this.isCatchingUp = false
+      return
+    }
+
+    if (lastBlock !== undefined && lastBlock.data.hash !== nextBlock.previousblockhash) {
+      await this._invalidate(lastBlock)
+    } else {
+      const queue = await this._addressQueue()
+      const _addresses = await this._getAddresses(nextBlock)
+      for (const a of _addresses) {
+        await queue.push(a)
+      }
+      await this.crawledBlockHashes.put({
+        partition: 'NONE',
+        id: nextBlock.hash,
+        sort: nextBlock.height,
+        data: nextBlock
+      })
+    }
+
+    return await this._catchUp()
+  }
+
+  private async _getAddresses (block: defid.Block<defid.Transaction>): Promise<string[]> {
+    const _addresses = []
+    for (const tx of block.tx) {
+      const addresses = await this.addressParser.parse(tx)
+      _addresses.push(...addresses)
+    }
+    return _addresses
+  }
+
+  private async _getBlock (height: number): Promise<defid.Block<defid.Transaction> | undefined> {
+    try {
+      const bh = await this.apiClient.blockchain.getBlockHash(height)
+      return await this.apiClient.blockchain.getBlock(bh, 2)
```

Comment:
![image](https://user-images.githubusercontent.com/44192990/155662586-1a9bb25e-6b86-4aa6-a8d1-e2688e2cf10b.png)

We have some linter rules in place since inherited since project beginning. I think either syntax has pros and cons, for I prefer still to return with await now.